compiler/rustc_passes/src/check_attr.rs

```diff
@@ -518,8 +542,12 @@ fn is_c_like_enum(item: &Item<'_>) -> bool {
 }
 
 fn check_mod_attrs(tcx: TyCtxt<'_>, module_def_id: LocalDefId) {
-    tcx.hir()
-        .visit_item_likes_in_module(module_def_id, &mut CheckAttrVisitor { tcx }.as_deep_visitor());
+    let mut visitor = CheckAttrVisitor { tcx };
+    tcx.hir().visit_item_likes_in_module(module_def_id, &mut visitor.as_deep_visitor());
+    let hir_id = tcx.hir().local_def_id_to_hir_id(module_def_id);
+    for attr in tcx.hir().krate_attrs() {
+        visitor.check_doc_alias(attr, hir_id, Target::Mod, true);
+    }
```

Comment:
`check_mod_attrs` is called for each module in a crate so isn't this going to emit the error for every module in the crate?

Reply:
Not the crate module apparently. I tested on this code:

```rust
#![feature(doc_alias)]
#![doc(alias = "test")]
```

Never saw an inner attribute. That's how I discovered that the krate module was passed down.

Reply:
So `check_mod_attrs` is called for every module in the crate including the crate root module but the problem is that it only checks items inside the module and not the module itself. This looks like a preexisting issue in rustc and I'm not sure how best to fix it. What is here though will check the crate attributes multiple times which is definitely not right.

I'm not actually convinced that we should disallow `#![doc(alias)]` for the crate root module as it is allowed on all other modules.

Reply:
I think it's intended this way.

> What is here though will check the crate attributes multiple times which is definitely not right.

What do you mean? As far as I can tell, it checks the crates' attributes once and that's it. Did I miss something?

> I'm not actually convinced that we should disallow `#![doc(alias)]` for the crate root module as it is allowed on all other modules.

In which case would we want to have a doc alias on the crate module itself?

Reply:
> I think it's intended this way.

rustc not checking crate level attributes is very much a bug causing #64734 and #69365 for example. I would like to see a proper fix for this.

> What do you mean? As far as I can tell, it checks the crates' attributes once and that's it. Did I miss something?

You can see from where `check_mod_attrs` is called that it is called once per module in the crate: https://github.com/rust-lang/rust/blob/da897dfb6daa268a965d2d73620316ba56020a19/compiler/rustc_interface/src/passes.rs#L812-L818 and https://github.com/rust-lang/rust/blob/445f34bb144c5b163163d8243e9845c644d3662d/src/librustdoc/core.rs#L514-L517

That means errors are emitted multiple times. This can be seen be compiling a crate with this PR applied like:

```rust
#![feature(doc_alias)]
#![doc(alias = "hello")]
mod foo {}
```

When `-Zdeduplicate-diagnostics=no` is passed to rustc or rustdoc the output is:

```
error: `#![doc(alias = "...")]` isn't allowed as a crate level attribute
 --> foo.rs:2:8
  |
2 | #![doc(alias = "hello")]
  |        ^^^^^^^^^^^^^^^

error: `#![doc(alias = "...")]` isn't allowed as a crate level attribute
 --> foo.rs:2:8
  |
2 | #![doc(alias = "hello")]
  |        ^^^^^^^^^^^^^^^

error: aborting due to 2 previous errors
```

Adding more modules causes the error to be repeated even more.

> In which case would we want to have a doc alias on the crate module itself?

I have no idea, but I also don't know why you would want doc aliases on any module. I'm fine with disallowing it because we can always allow it in the future if someone has a use case.

Reply:
Actually, we need them for modules because of primitives. :)